module1-introduction-to-backend/r3.4-api-best-practices/README.md

```diff
@@ -0,0 +1,232 @@
+]
+
+app.use(bodyParser.json());
+
+app.get('/employees', (req, res) => {
+  const { firstName, lastName, age } = req.query;
+  let results = [...employees];
+  if (firstName) {
+    results = results.filter(r => r.firstName === firstName);
+  }
+
+  if (lastName) {
+    results = results.filter(r => r.lastName === lastName);
+  }
+
+  if (age) {
+    results = results.filter(r => +r.age === +age);
+  }
+  res.json(results);
+});
+
+app.listen(3000, () => console.log('server started'));
+```
+
+7. **Maintain good security practices**
+Most communication between client and server should be private since we often send and receive private information. Therefore, using SSL/TLS for security is a must. A SSL certificate isn’t too difficult to load onto a server and the cost is free or very low. There’s no reason not to make our REST APIs communicate over secure channels instead of in the open.
+
+People shouldn’t be able to access more information that they requested. For example, a normal user shouldn’t be able to access information of another user. They also shouldn’t be able to access data of admins. To enforce the principle of least privilege, we need to add role checks either for a single role, or have more granular roles for each user.
+
+If we choose to group users into a few roles, then the roles should have the permissions that cover all they need and no more. If we have more granular permissions for each feature that users have access to, then we have to make sure that admins can add and remove those features from each user accordingly. Also, we need to add some preset roles that can be applied to a group users so that we don’t have to do that for every user manually.
+
+8. **Cache data to improve performance**
+We can add caching to return data from the local memory cache instead of querying the database to get the data every time we want to retrieve some data that users request. The good thing about caching is that users can get data faster. However, the data that users get may be outdated. This may also lead to issues when debugging in production environments when something goes wrong as we keep seeing old data.
+
+There are many kinds of caching solutions like Redis, in-memory caching, and more. We can change the way data is cached as our needs change.
+
+9. **Versioning our APIs**
+We should have different versions of API if we’re making any changes to them that may break clients. The versioning can be done according to semantic version (for example, 2.0.6 to indicate major version 2 and the sixth patch) like most apps do nowadays.
+
+This way, we can gradually phase out old endpoints instead of forcing everyone to move to the new API at the same time. The v1 endpoint can stay active for people who don’t want to change, while the v2, with its shiny new features, can serve those who are ready to upgrade. This is especially important if our API is public. We should version them so that we won’t break third party apps that use our APIs.
```

Comment:
the concrete example i would give here is mobile apps, and mobile apps its usually hard to force users to update, or people may be on old versions, therefore you could still have clients that expect older versions